Toolkit/ArcGISToolkit/TimeSlider.swift

```diff
@@ -0,0 +1,2603 @@
+                currentExtentEndTimeLabel.isHidden = true
+            }
+            else if endTimeLabelX + endTimeLabelSize.width > bounds.maxX - labelSidePadding {
+                endTimeLabelX = bounds.maxX - endTimeLabelSize.width - labelSidePadding
+                currentExtentEndTimeLabel.isHidden = true
+            }
+            else if endTimeLabelX < currentExtentStartTimeLabel.frame.origin.x + currentExtentStartTimeLabel.frame.width {
+                endTimeLabelX = currentExtentStartTimeLabel.frame.origin.x + currentExtentStartTimeLabel.frame.width + paddingBetweenLabels
+            }
+            
+            let thumbEndTimeLabelY = upperThumbLayer.frame.minY - currentExtentEndTimeLabel.frame.height - labelPadding
+            let tickLayerEndTimeLabelY = tickMarkLayer.frame.minY - currentExtentEndTimeLabel.frame.height - labelPadding
+            let endTimeLabelY = min(thumbEndTimeLabelY, tickLayerEndTimeLabelY)
+            currentExtentEndTimeLabel.frame = CGRect(x: endTimeLabelX, y: endTimeLabelY, width: endTimeLabelSize.width, height: endTimeLabelSize.height)
+        }
+        else {
+            currentExtentEndTimeLabel.string = ""
+            currentExtentEndTimeLabel.isHidden = true
+        }
+        
+        // Commit the transaction
+        CATransaction.commit()
+    }
+    
+    private func positionTickMarks() {
+        //
+        // Bail out if time steps are not available
+        guard let timeSteps = timeSteps, !timeSteps.isEmpty, isSliderVisible else {
+            return
+        }
+        
+        // Create tick marks based on
+        // time steps and it's calculated
+        // position on the slider
+        let tms = timeSteps.map { timeStep -> TickMark in
+            let tickX = CGFloat(position(for: timeStep.timeIntervalSince1970))
+            return TickMark(originX: tickX, value: timeStep)
+        }
+        
+        // If label mode is ticks then we need to
+        // find out the major/minor ticks and
+        // add labels for major ticks
+        if labelMode == .ticks {
+            var majorTickInterval = 2
+            var doMajorTickLabelsCollide = false
+            var firstMajorTickIndex = 0
+            let tickCount = tms.count
+            
+            // Calculate the largest number of ticks to allow between major ticks. This prevents scenarios where
+            // there are two major ticks placed undesirably close to the end.
+            let maxMajorTickInterval = Int(ceil(Double(tickCount / 2)))
+            
+            // If maxMajorTickInterval is not greater than majorTickInterval
+            // then we can not do further calculation
+            if maxMajorTickInterval >= majorTickInterval {
+                //
+                // Calculate the number of ticks between each major tick and the index of the first major tick
+                for i in majorTickInterval..<maxMajorTickInterval  {
+                    let prospectiveInterv
```

Comment:
@njarecha Could you please make this a `struct` instead of a `class`?

Reply:
I'm not sure which line this comment is for. Could you please let me know. I can't see this comment in Files changed page.

Reply:
@njarecha This is on line 2231: `class TickMark` -> `struct TickMark`.

Reply:
Objects of this class is being stored in the Array I expect them to be stored as reference. The `struct` will store them as copy and it does not work as expected. Hence, it should be `class`.

Reply:
Ah! I see what you mean. Okay, I suppose that's fine. 

It would be nice to make `TickMark` a value type. There would be a performance benefit and it would allow you to observe when `tickMarks` changes. For example, you could refactor updating tick mark labels out of `positionTickMarks()`:
```swift
fileprivate var tickMarks = [TickMark]() {
    didSet {
        guard tickMarks != oldValue else { return }
        updateTickMarkLabels()
    }
}
```